server/e2e_test.go

```diff
@@ -227,24 +209,25 @@ func (c *testClient) Connect() {
 	c.inbox = newTestInbox(c.conn)
 
 	// handle the server side of the connection
-	c.h.tracker.mutex.Lock()
-	defer c.h.tracker.mutex.Unlock()
 	go handleConnection(c.remoteConn, c.h.packets, c.h.tracker)
 }
 
-// Disconnect simulates the client dropping the connection
+// Disconnect simulates the client dropping the connection and confirms that
+// the server is aware of the disconnect.
 func (c *testClient) Disconnect() {
 	// client side
 	if c.conn != nil {
 		if err := c.conn.Close(); err != nil {
-			panic(err)
+			c.h.t.Fatal(err)
 		}
 	}
+	c.h.WaitNotConnected(c)
```

Comment:
This is another important functional change. Again encapsulating all state confirmations within the action so that when it returns, the server should be in a known state and not require additional waiting.

Reply:
+1